### PR TITLE
New Widget.detach() is a user function called at detach time

### PR DESCRIPTION
**Pull request by swankjesse**
This is invoked when a view will never be operated by the presenter. It usually happens when a view is removed from the view tree, but may also happen later if the view or one of its parent views is pooled.

Implementations should do what they can to make Kotlin objects eligible for immediate garbage collection. In particular a useful thing to do is to remove references from reference-counted Swift objects to Kotlin objects. That way the GC can reclaim the memory sooner, and we may reduce the chance of encountering KT-68769.

https://youtrack.jetbrains.com/issue/KT-68769/EXCBADACCESS-Attempted-to-dereference-garbage-pointer-0x28

---

- [x] `CHANGELOG.md`'s "Unreleased" section has been updated, if applicable.

### CODE REVIEW

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt [hunk 1]
```diff
@@ -68,16 +68,31 @@ import platform.darwin.NSObject
...
+
+  internal class UIViewLazyListTableView :
+    UITableView(
+      CGRectZero.readValue(),
+      UITableViewStyle.UITableViewStylePlain,
+    ) {
+    var scrollProcessor: UIViewLazyListScrollProcessor? = null
+
     override fun setContentOffset(contentOffset: CValue<CGPoint>, animated: Boolean) {
+      val scrollProcessor = this.scrollProcessor ?: return // Detached.
```
Comment (dnagler) [inline redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:89]: Interesting idea!

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt [hunk 2]
```diff
@@ -260,6 +314,20 @@ internal open class UIViewLazyList :
...
+    // Reverse what we set up in init {}. This is strictly unnecessary but should maximize the
+    // number of Kotlin objects that are immediately eligible for garbage collection.
+    scrollProcessor.tableView = null
+    scrollProcessor.updateProcessor = null
+    updateProcessor.tableView = null
+    dataSource.updateProcessor = null
+    tableViewDelegate.tableView = null
+    tableViewDelegate.scrollProcessor = null
+    tableView.scrollProcessor = null
+    // tableView.dataSource = null // Don't clear the data source, it's weakly held anyway.
```
Comment (dnagler) [inline redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt]: nit: might as well delete the commented-out code here (the comment is useful though!)

File: redwood-testing/src/commonTest/kotlin/app/cash/redwood/testing/WidgetDetachingTest.kt [hunk 3]
```diff
@@ -56,14 +57,20 @@ class WidgetDetachingTest {
...
       awaitSnapshot()
-      val thrown = assertFailsWith<IllegalStateException> {
+      val thrownOnAccessWidget = assertFailsWith<IllegalStateException> {
         protocolText.widget
       }
-      assertThat(thrown).hasMessage("detached")
+      assertThat(thrownOnAccessWidget).hasMessage("detached")
+
+      val thrownOnMutateWidget = assertFailsWith<IllegalStateException> {
+        protocolTextWidget?.text("goodbye")
```
Comment (dnagler) [inline redwood-testing/src/commonTest/kotlin/app/cash/redwood/testing/WidgetDetachingTest.kt:68]: lol

File: redwood-tooling-codegen/src/main/kotlin/app/cash/redwood/tooling/codegen/protocolHostGeneration.kt [hunk 4]
```diff
@@ -506,6 +506,7 @@ internal fun generateProtocolNode(
                 }
               }
             }
+            .addStatement("_widget?.detach()")
```
Comment (dnagler) [inline redwood-tooling-codegen/src/main/kotlin/app/cash/redwood/tooling/codegen/protocolHostGeneration.kt:509]: nice

**Review by dnagler — APPROVED** [review]

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt [hunk 5]
```diff
@@ -68,16 +68,31 @@ import platform.darwin.NSObject
 internal open class UIViewLazyList :
```
Comment (swankjesse) [inline redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:68]: There is a lot of Swift and Kotlin stuff happening in this file. Rather than being precise and eliminating just the references from Swift objects to Kotlin ones, I just cut down the reference graph aggressively.

There’s still a few places where Swift objects will hold Kotlin objects...

The cells in the table will be instances of `LazyListContainerCell`. This is a UIView subclass implemented in Kotlin, and honestly I don’t have intuition on how reachability works with such things. Each cell has a `Binding` that I might also need to clear.

The table has a datasource and a delegate, both of which are weakly held. I also don’t have much intuition on how a Swift weak references to a Kotlin object is implemented.

File: redwood-widget/src/commonMain/kotlin/app/cash/redwood/widget/Widget.kt [hunk 6]
```diff
@@ -38,6 +38,16 @@ public interface Widget<W : Any> {
    */
   public var modifier: Modifier
 
+  /**
+   * Clear event listeners **without** triggering an update the displayed UI.
```
Comment (JakeWharton) [inline redwood-widget/src/commonMain/kotlin/app/cash/redwood/widget/Widget.kt]:
```suggestion
   * Clear event listeners **without** triggering an update to the displayed UI.
```
or "on"?

File: redwood-widget/src/commonMain/kotlin/app/cash/redwood/widget/Widget.kt [hunk 7]
```diff
@@ -38,6 +38,16 @@ public interface Widget<W : Any> {
...
+   *
+   * After this is called there will be no further updates to properties or children.
+   *
+   * This function is not recursive.
+   */
+  public fun detach() {
```
Comment (JakeWharton) [inline redwood-widget/src/commonMain/kotlin/app/cash/redwood/widget/Widget.kt:48]: This doesn't appear to ever be called if you aren't using the protocol.

We're also going to force every binding written in Swift to override this.

**Review by JakeWharton — APPROVED** [review]

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt (hunk 2 above)
Comment (swankjesse) [reply redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt]: Fixed!

File: redwood-widget/src/commonMain/kotlin/app/cash/redwood/widget/Widget.kt (hunk 6 above)
Comment (swankjesse) [reply redwood-widget/src/commonMain/kotlin/app/cash/redwood/widget/Widget.kt]: Fixed!

File: redwood-widget/src/commonMain/kotlin/app/cash/redwood/widget/Widget.kt (hunk 7 above)
Comment (swankjesse) [reply redwood-widget/src/commonMain/kotlin/app/cash/redwood/widget/Widget.kt:48]: Punting on the non-protocol dispose(). If we’re lucky we can delete the dispose() function soon and never implement it there. https://github.com/cashapp/redwood/issues/2084

Agreed on forcing Swift bindings to implement this. I think it’s fine.
